lib/chan.ml

```diff
@@ -2,73 +2,104 @@ type 'a contents =
   | Empty of {receivers: ('a option ref * Domain.id) Fun_queue.t}
   | NotEmpty of {senders: ('a * Domain.id) Fun_queue.t; messages: 'a Fun_queue.t}
 
-type 'a t = {buffer_size: int; contents: 'a contents Atomic.t}
+type 'a t = {buffer_size: int option; contents: 'a contents Atomic.t}
```

Comment:
> if instead of length messages = buffer_size one checks length messages + 1 > buffer_size

I don't understand this. We only check if `buffer_size = Some (length messages)`. Even if `buffer_size` were `max_int`, this shouldn't cause an overflow IIUC.

Reply:
I also tried making `buffer_size` a type that distinguishes 0, bounded and unbounded lengths. What might be better is to structure the channel such that 0 length channels will have a separate variant without the `messages` field. Given that 0 length channels should not have any messages, such a data type will make those cases unrepresentable. I'd probably save this change for later.

Reply:
(Re. overflow: with the current code if you use `max_int` you do not *really* get unbounded channels, as you stop pushing to it when you read `max_int` messages. Unless I'm mistaken, the test I mentioned ensures that the "buffer is full" condition never actually fires in the `max_int` case, by, erm, abusing overflows in a horrible way. Let's forget about this conversation.)

Thanks for the clarification. (I might try refactoring the refactoring you suggest to get a chance to understand the implementation better.)

Reply:
>  Let's forget about this conversation.

:-) 

> Thanks for the clarification. (I might try refactoring the refactoring you suggest to get a chance to understand the implementation better.)

Wonderful. 👍